src/unix/linux.c

```diff
@@ -1341,47 +1341,139 @@ static uint64_t uv__read_uint64(const char* filename) {
 }
 
 
-/* This might return 0 if there was a problem getting the memory limit from
- * cgroups. This is OK because a return value of 0 signifies that the memory
- * limit is unknown.
- */
-static uint64_t uv__get_constrained_memory_fallback(void) {
-  return uv__read_uint64("/sys/fs/cgroup/memory/memory.limit_in_bytes");
-}
+static void uv__get_cgroup1_memory_limits(char* buf, uint64_t* high,
+                                          uint64_t* max) {
+  char filename[4097];
+  char* p;
+  char* p_end;
+
+  /* Find out the path of the memory controller. */
+  p = strchr(buf, ':');
+  while (p != NULL && memcmp(p, ":memory:", 8)) {
+    p = strchr(p, '\n');
+    if (p != NULL)
+      p = strchr(p, ':');
+  }
+
+  if (p != NULL) {
+    p = p + 8;
+    p_end = strchr(p, '\n');
+    if (p_end != NULL)
+      *p_end = '\0';
+
+    /* If the memory controller is mounted, check limits there. */
+    snprintf(filename, sizeof(filename), "/sys/fs/cgroup/memory/%s/", p);
+    if (0 == access(filename, F_OK)) {
+      /* Calls to uv__read_uint64 might return 0 if there was a problem. This is
+       * OK because a return value of 0 signifies that the limit is unknown.
+       */
+      snprintf(filename, sizeof(filename),
+              "/sys/fs/cgroup/memory/%s/memory.soft_limit_in_bytes", p);
+      *high = uv__read_uint64(filename);
 
+      snprintf(filename, sizeof(filename),
+              "/sys/fs/cgroup/memory/%s/memory.limit_in_bytes", p);
+      *max = uv__read_uint64(filename);
 
-uint64_t uv_get_constrained_memory(void) {
+      return;
+    }
+  }
+
+  /* Fall back to the global limits of the memory controller. */
+  *high = uv__read_uint64("/sys/fs/cgroup/memory/memory.soft_limit_in_bytes");
+  *max = uv__read_uint64("/sys/fs/cgroup/memory/memory.limit_in_bytes");
+}
+
+static void uv__get_cgroup2_memory_limits(char* buf, uint64_t* high,
+                                          uint64_t* max) {
   char filename[4097];
-  char buf[1024];
+  char* p;
+
+  p = strchr(buf, '\n');
+  if (p != NULL)
+    *p = '\0';
+
+  p = buf + 4;
```

Comment:
There's a bug here. I think you think you're skipping over `0::/` but `p` is either null or pointing to `\n` at this point.

Reply:
I think this works as intended, as it does `buf + 4`, not `p + 4`. But I'll change it to the non-mutating version as you suggested above.